src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -0,0 +1,443 @@
+                        k
+                    );
+                }
+            }
+            prost::encoding::encode_key(tag, WireType::LengthDelimited, result);
+            prost::encoding::encode_varint(struct_buf.len() as u64, result);
+            result.append(&mut struct_buf);
+        }
+        TableType::Bytes => {
+            prost::encoding::bytes::encode(
+                tag,
+                &Vec::from(val.as_bytes().ok_or_else(|| {
+                    ErrorKind::BigQueryTypeMismatch("bytes", format!("{:?}", val.value_type()))
+                })?),
+                result,
+            );
+        }
+        TableType::Json => {
+            warn!("Found a field of type JSON, this is not supported, ignoring.");
+        }
+        TableType::Interval => {
+            warn!("Found a field of type Interval, this is not supported, ignoring.");
+        }
+
+        TableType::Unspecified => {
+            warn!("Found a field of unspecified type - ignoring.");
+        }
+    }
+
+    Ok(())
+}
+
+impl JsonToProtobufMapping {
+    pub fn new(vec: &Vec<TableFieldSchema>) -> Self {
+        let descriptor = map_field("table", vec);
+
+        Self {
+            descriptor: descriptor.0,
+            fields: descriptor.1,
+        }
+    }
+
+    pub fn map(&self, value: &Value) -> Result<Vec<u8>> {
+        let mut result = vec![];
+        if let Some(obj) = value.as_object() {
+            for (key, val) in obj {
+                if let Some(field) = self.fields.get(&key.to_string()) {
+                    encode_field(val, field, &mut result)?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    pub fn descriptor(&self) -> &DescriptorProto {
+        &self.descriptor
+    }
+}
+impl GbqSink {
+    pub async fn new(config: Config) -> Result<Self> {
+        Ok(Self {
+            client: None,
+            write_stream: None,
+            mapping: None,
+            config,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for GbqSink {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        _serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self
+            .client
+            .as_mut()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                "The client is not connected",
+            ))?;
+        let write_stream =
+            self.write_stream
+                .as_ref()
+                .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                    "The write stream is not available",
+                ))?;
+        let mapping = self
+            .mapping
+            .as_ref()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                "The mapping is not available",
+            ))?;
+
+        let request = AppendRowsRequest {
```

Comment:
We will need to handle batching here, right now a batched (coming from the batch operator) request would fail the encoding. I think ES / HTTP have some good examples on how to best do this. It'd also allow multiple rows to be inserted at once via batching wich would be a nice feature.